modules/dnn/src/layers/max_unpooling_layer.cpp

```diff
@@ -71,9 +78,63 @@ class MaxUnpoolLayerImpl CV_FINAL : public MaxUnpoolLayer
         inputs_arr.getMatVector(inputs);
         outputs_arr.getMatVector(outputs);
 
-        CV_Assert(inputs.size() == 2);
-        Mat& input = inputs[0];
-        Mat& indices = inputs[1];
+        Mat input;
+        Mat indices;
+        if (inputs.size() == 2)
+        {
+            input = inputs[0];
+            indices = inputs[1];
+        }
+        else if (inputs.size() == 3)
+        {
+            input = inputs[2];
+
+            int kernel_area = poolKernel.height * poolKernel.width;
+            Mat maxpoolintp = inputs[0];
+            Mat maxpooloutp = inputs[1];
+            
+            int sz[] = { 1, input.size[1], input.size[2], input.size[3] };
+            indices.create(4, sz, CV_32F);
+
+            std::vector<Mat> in_planes;
+            std::vector<Mat> out_planes;
+            std::vector<Mat> indice_arr;
+            for (int i_c = 0; i_c < input.size[1]; i_c++)
+            {
+                in_planes.push_back(getPlane(maxpoolintp, 0, i_c));
+                out_planes.push_back(getPlane(maxpooloutp, 0, i_c));
+                indice_arr.push_back(getPlane(indices, 0, i_c));
+            }
+
+            for (int i_h = 0; i_h < maxpoolintp.size[2]; i_h += poolKernel.height)
+            {
+                if (i_h / poolKernel.height < maxpooloutp.size[2])
+                {
+                    for (int i_w = 0; i_w < maxpoolintp.size[3]; i_w += poolKernel.width)
+                    {
+                        if (i_w / poolKernel.width < maxpooloutp.size[3])
+                        {
+                            for (int i_c = 0; i_c < input.size[1]; i_c++)
+                            {
+                                int index = 0;
+                                for (; index < kernel_area; ++index)
+                                {
+                                    auto out = out_planes[i_c].at<float>(i_h / poolKernel.height, i_w / poolKernel.width);
+                                    auto in = in_planes[i_c].at<float>(i_h + index / poolKernel.width, i_w + index % poolKernel.width);
+                                    if (abs(out - in) < 0.00001)
+                                    {
+                                        indice_arr[i_c].at<float>(i_h / poolKernel.height, i_w / poolKernel.width) = (i_h + index / poolKernel.width) * maxpoolintp.size[3] + i_w + index % poolKernel.width;
+                                        break;
+                                    }
+                                }
+                                if (index == kernel_area)
+                                    indice_arr[i_c].at<float>(i_h / poolKernel.height, i_w / poolKernel.width) = -1;
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Let's try to trigger existing implementation. Usage of hardcoded thresholds is too unstable.

Out poolling layer can produce indices of maximum values if it has consumers to it's second pin (the first one are values). So you need to use `connect` to connect Unpooling layer to it's input (some layer) and the 1st pin of pooling layer. To be sure that you connected everything right I'd like to suggest to add one more layer in you network (i.e. 1x1 convolution):

```
input ---+---> pooling -> convolution -> unpooling
         |                                    ^
         +------------------------------------+
```

Reply:
OK, I wil try it.

Reply:
How can i connect to maxpool's maxid result?  
when i connect maxpool like this:
`connect(layer_id, dstNet, parsePin(pool_name), id, 0);`
the maxpool's output size will be set to 1 and computeMaxIdx be set to false.